pkg/operator/quorumguardcontroller/quorumguardcontroller.go

```diff
@@ -0,0 +1,243 @@
+			Type:    "QuorumGuardControllerDegraded",
+			Status:  operatorv1.ConditionTrue,
+			Reason:  "Error",
+			Message: err.Error(),
+		}))
+		if updateErr != nil {
+			syncCtx.Recorder().Warning("QuorumGuardControllerUpdatingStatus", updateErr.Error())
+		}
+		return err
+	}
+
+	_, _, updateErr := v1helpers.UpdateStatus(c.operatorClient,
+		v1helpers.UpdateConditionFn(operatorv1.OperatorCondition{
+			Type:   "QuorumGuardControllerDegraded",
+			Status: operatorv1.ConditionFalse,
+			Reason: "AsExpected",
+		}))
+	return updateErr
+}
+
+func (c *QuorumGuardController) ensureEtcdGuardDeployment(ctx context.Context, recorder events.Recorder) error {
+	haMode, err := c.isInHAMode(ctx)
+	if err != nil {
+		klog.Infof("Failed to validate ha mode %v ", err)
+		return err
+	}
+	if !haMode {
+		return nil
+	}
+
+	replicaCount, err := c.getMastersReplicaCount(ctx)
+	if err != nil {
+		return err
+	}
+
+	if err := c.ensureEtcdGuardDeployment2(ctx, replicaCount, recorder); err != nil {
+		return err
+	}
+
+	if err := c.ensureEtcdGuardPdbDeployment(ctx, recorder); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (c *QuorumGuardController) ensureEtcdGuardDeployment2(ctx context.Context, replicaCount int32, recorder events.Recorder) error {
+	_, err := c.kubeClient.AppsV1().Deployments(operatorclient.TargetNamespace).Get(ctx,
+		EtcdGuardDeploymentName, metav1.GetOptions{})
+	switch {
+	case apierrors.IsNotFound(err):
+		recorder.Eventf("NoEtcdGuardDeployment", "%s was not found, creating one", EtcdGuardDeploymentName)
+		return c.applyDeployment(replicaCount, recorder)
+	case err != nil:
+		klog.Warningf("failed to get %s deployment err %v", EtcdGuardDeploymentName, err.Error())
+		return err
+	default:
+		return nil
+	}
+}
+
+func (c *QuorumGuardController) ensureEtcdGuardPdbDeployment(ctx context.Context, recorder events.Recorder) error {
+	_, err := c.kubeClient.PolicyV1beta1().PodDisruptionBudgets(operatorclient.TargetNamespace).Get(ctx,
+		EtcdGuardDeploymentName, metav1.GetOptions{})
+	switch {
+	case apierrors.IsNotFound(err):
+		recorder.Eventf("NoEtcdGuardDeployment", "%s was not found, creating one", EtcdGuardDeploymentName)
+		return c.applyPdbDeployment(ctx, recorder)
+	case err != nil:
+		klog.Warningf("failed to get %s deployment err %v", EtcdGuardDeploymentName, err.Error())
+		return err
+	default:
+		return nil
+	}
+}
+
+// Applying etcd guard deployment
+func (c *QuorumGuardController) applyDeployment(replicaCount int32, recorder events.Recorder) error {
+	klog.Infof("Going to create new %s deployment with replica count %d", EtcdGuardDeploymentName, replicaCount)
+
+	deployment := resourceread.ReadDeploymentV1OrDie(etcd_assets.MustAsset("etcd/quorumguard-deployment.yaml"))
+	deployment.Spec.Replicas = &replicaCount
+	newDeployment, _, err := resourceapply.ApplyDeployment(c.kubeClient.AppsV1(), recorder, deployment, -1)
+	klog.Infof("New deployment is %v", newDeployment)
+	return err
```

Comment:
nit can we use regular golang patterns here if err != nil ... then return nil?

Reply:
we can but the idea was that we don't need one more "if"